core/dblookupext/eventsHashesByTxHashIndex.go

```diff
@@ -0,0 +1,133 @@
+
+package dblookupext
+
+import (
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/data/receipt"
+	"github.com/ElrondNetwork/elrond-go/data/smartContractResult"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/ElrondNetwork/elrond-go/storage"
+)
+
+type eventsHashesByTxHash struct {
+	marshalizer marshal.Marshalizer
+	storer      storage.Storer
+}
+
+func newEventsHashesByTxHash(storer storage.Storer, marshalizer marshal.Marshalizer) *eventsHashesByTxHash {
+	return &eventsHashesByTxHash{
+		marshalizer: marshalizer,
+		storer:      storer,
+	}
+}
+
+func (eht *eventsHashesByTxHash) saveResultsHashes(epoch uint32, scResults, receipts map[string]data.TransactionHandler) error {
+	resultsHashes := eht.groupTransactionOutcomesByTransactionHash(epoch, scResults, receipts)
+	for txHash, resultHashes := range resultsHashes {
+		resultHashesBytes, err := eht.marshalizer.Marshal(resultHashes)
+		if err != nil {
+			continue
+		}
+
+		err = eht.storer.Put([]byte(txHash), resultHashesBytes)
+		if err != nil {
+			log.Warn("saveResultsHashes() cannot save resultHashesByte",
+				"error", err.Error())
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (eht *eventsHashesByTxHash) groupTransactionOutcomesByTransactionHash(
+	epoch uint32,
+	scResults,
+	receipts map[string]data.TransactionHandler,
+) map[string]*ResultsHashesByTxHash {
+	resultsHashesMap := eht.groupSmartContractResults(epoch, scResults)
+
+	for receiptHash, receiptHandler := range receipts {
+		rec, ok := receiptHandler.(*receipt.Receipt)
+		if !ok {
+			log.Error("groupTransactionOutcomesByTransactionHash() cannot cast TransactionHandler to Receipt")
+			continue
+		}
+
+		originalTxHash := string(rec.TxHash)
+		resultsHashesMap[originalTxHash] = &ResultsHashesByTxHash{
+			ReceiptsHash: []byte(receiptHash),
+		}
+	}
+
+	return resultsHashesMap
+}
+
+func (eht *eventsHashesByTxHash) groupSmartContractResults(
+	epoch uint32,
+	scrResults map[string]data.TransactionHandler,
+) map[string]*ResultsHashesByTxHash {
+	resultsHashesMap := make(map[string]*ResultsHashesByTxHash, 0)
+	for scrHash, scrHandler := range scrResults {
+		scrResult, ok := scrHandler.(*smartContractResult.SmartContractResult)
+		if !ok {
+			log.Error("groupSmartContractResults() cannot cast TransactionHandler to SmartContractResult")
+			continue
+		}
+
+		originalTxHash := string(scrResult.OriginalTxHash)
+
+		if _, ok := resultsHashesMap[originalTxHash]; !ok {
+			resultsHashesMap[originalTxHash] = &ResultsHashesByTxHash{
+				ScResultsHashesAndEpoch: []*ScResultsHashesAndEpoch{
+					{
+						Epoch: epoch,
+					},
+				},
+			}
+		}
+		scResultsHashesAndEpoch := resultsHashesMap[originalTxHash].ScResultsHashesAndEpoch[0]
+		scResultsHashesAndEpoch.ScResultsHashes = append(scResultsHashesAndEpoch.ScResultsHashes, []byte(scrHash))
+	}
+
+	results := eht.mergeRecordsFromStorageIfExists(resultsHashesMap)
```

Comment:
Here you call `mergeFromStorage` in the group function. Can you perhaps call this merge function from `saveResultsHashes` instead? The group function should not read from storage, not its business.

Reply:
done.